### ISSUE TITLE
Exception RuntimeError: RuntimeError('main thread is not in main loop',)

### ISSUE THREAD

**Issue by tvrbanec** [joblib/joblib#807]
In the main program I call modules (five times). In each module I use joblib.Parallel in the same way. If I call only one of these modules from the main program, everything is doing great. If in the main program I choose the option to execute the five modules one by one, in the third of them I get the following error:

```python-traceback
Exception RuntimeError: RuntimeError('main thread is not in main loop',) in <bound method PhotoImage.__del__ of <Tkinter.PhotoImage instance at 0x7f72ac673c68>> ignored
Tcl_AsyncDelete: async handler deleted by the wrong thread
Aborted
```

When only that third subprogram (module) is called, everything is executed without error.
joblib-0.13.0, Debian

**Comment by tomMoral** [comment]
Thanks for reporting this.

With the information you give us, it is very hard to know what might go wrong. It seems to be a bad interaction with `asyncio`. Could you provide a [MCVE](http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) for this issue?

Also, what do you mean by modules? Is that coroutine or is it something else?

**Comment by tvrbanec** [comment]
Coroutine.
I can't provide MCVE - it would take too much time.

This is the end of error report:
/usr/local/lib/python2.7/dist-packages/gensim/models/word2vec.py in init_sims(self=<gensim.models.word2vec.Word2Vec object>, replace=True)
   1158         See :meth:`~gensim.models.keyedvectors.Word2VecKeyedVectors.init_sims`.
   1159 
   1160         """
   1161         if replace and hasattr(self.trainables, 'syn1'):
   1162             del self.trainables.syn1
-> 1163         return self.wv.init_sims(replace)
        self.wv.init_sims = <bound method Word2VecKeyedVectors.init_sims of ...models.keyedvectors.Word2VecKeyedVectors object>>
        replace = True
   1164 
   1165     def reset_from(self, other_model):
   1166         """Borrow shareable pre-built structures from `other_model` and reset hidden layer weights.
   1167 

...........................................................................
/usr/local/lib/python2.7/dist-packages/gensim/models/keyedvectors.py in init_sims(self=<gensim.models.keyedvectors.Word2VecKeyedVectors object>, replace=True)
   1358         """
   1359         if getattr(self, 'vectors_norm', None) is None or replace:
   1360             logger.info("precomputing L2-norms of word weight vectors")
   1361             if replace:
   1362                 for i in xrange(self.vectors.shape[0]):
-> 1363                     self.vectors[i, :] /= sqrt((self.vectors[i, :] ** 2).sum(-1))
        self.vectors = memmap([[ 0.24386726, -0.15418427, -0.01848165, ...       -0.03422905,  0.04567238]], dtype=float32)
        i = 0
        i.sum = undefined
   1364                 self.vectors_norm = self.vectors
   1365             else:
   1366                 self.vectors_norm = (self.vectors / sqrt((self.vectors ** 2).sum(-1))[..., newaxis]).astype(REAL)
   1367 

ValueError: output array is read-only

**Comment by tomMoral** [comment]
It looks like you have an issue with the `memmaps`. The `numpy` arrays which are passed as argument of a function in `joblib` are usually readonly. If this is the case, you could copy the arry before modifying it.

I cannot be more specific without a MVCE, as I do not see where you are calling `Parallel`. At least, the code of the function which makes the call to `Parallel` and the full traceback are needed to help you.

**Comment by tvrbanec** [comment]
I think the problem might be in tkinter: Tkinter is not quite thread safe.
https://stackoverflow.com/questions/14694408/runtimeerror-main-thread-is-not-in-main-loop.

**Comment by tvrbanec** [comment]
I really do not know exactly how, but the problem is solved. I've been trying a lot, changed a lot. Perhaps the pre_dispatch parameter was set too high, so it was out of reach.

**Comment by tvrbanec** [comment]
@tomMoral 
Thank you!

**Comment by Naereen** [comment]
Hi @tomMoral @tvrbanec,
I'm currently suffering from the same problem, and I have no idea how to solve it… (see https://github.com/SMPyBandits/SMPyBandits/issues/159).

The context might be different from @tvrbanec's context. I run some mathematical simulations, with n algorithms (eg n=25), and for each algorithm I run for instance 480 random repetitions of the same experiment. The part I run in parallel is these repetitions, for instance on 12 cores. I was using successfully joblib since the last two years!
Everytime I now get an error:
```python
Exception ignored in: <bound method Image.__del__ of <tkinter.PhotoImage objec
t at 0x7f5eead78f28>>
Traceback (most recent call last):
  File "/usr/lib/python3.5/tkinter/__init__.py", line 3359, in __del__
    self.tk.call('image', 'delete', self.name)
RuntimeError: main thread is not in main loop
Tcl_AsyncDelete: async handler deleted by the wrong thread
```

- I have tried increasing the `pre_dispatch` argument of `joblib.Parallel` to `pre_dispatch='10*n_jobs'`, it's not changing anything.
- I am not sure when/where any Tkinter object is/was created, but most probably it's because I start the simulation with an initial `matplotlib` plot (that's created and saved but *not shown*, hence my misunderstanding regarding the origin of this Tkinter object). If I disable this first plot, the bug no longer appear (see https://github.com/SMPyBandits/SMPyBandits/blob/master/SMPyBandits/main.py#L166).

Any idea of directions I could try? Thanks in advance!

**Comment by tomMoral** [comment]
@Naereen It seems like a bad interaction with `tkinter` library.

It is hard to debug without seeing the code which is run by the parallel loop.
Could you give a way to [reproduce your error](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-report) (_e.g._ a script to launch using SMPyBandits) in a separate issue?

**Comment by Naereen** [comment]
My diagnosis seems correct, I won't need more, I just changed back and no longer open any `matplotlib` plots before being completely done with the `Parallel`-code.
It's not worth my time trying to find a better solution. Thanks again for your reply!

**Comment by tomMoral** [comment]
Thanks for reporting back.
